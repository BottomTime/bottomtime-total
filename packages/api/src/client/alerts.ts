import {
  AlertSchema,
  ApiList,
  CreateOrUpdateAlertParamsDTO,
  ListAlertsParamsDTO,
  ListAlertsResponseSchema,
} from '../types';
import { Alert } from './alert';
import { Fetcher } from './fetcher';

export class AlertsApiClient {
  constructor(private readonly apiClient: Fetcher) {}

  async listAlerts(options?: ListAlertsParamsDTO): Promise<ApiList<Alert>> {
    const { data: response } = await this.apiClient.get(
      '/api/alerts',
      options,
      ListAlertsResponseSchema,
    );

    return {
      data: response.data.map((alert) => new Alert(this.apiClient, alert)),
      totalCount: response.totalCount,
    };
  }

  async getAlert(alertId: string): Promise<Alert> {
    const { data } = await this.apiClient.get(
      `/api/alerts/${alertId}`,
      undefined,
      AlertSchema,
    );
    return new Alert(this.apiClient, data);
  }

  async createAlert(options: CreateOrUpdateAlertParamsDTO): Promise<Alert> {
    const { data } = await this.apiClient.post(
      '/api/alerts',
      options,
      AlertSchema,
    );
    return new Alert(this.apiClient, data);
  }

  wrapDTO(data: unknown): Alert {
    return new Alert(this.apiClient, AlertSchema.parse(data));
  }
}
