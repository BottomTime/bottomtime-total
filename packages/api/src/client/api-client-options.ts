import { Fetcher } from './fetcher';

export type ApiClientOptions = {
  authToken?: string;
  baseURL?: string;
  edgeAuthToken?: string;
  fetcher?: Fetcher;
};
