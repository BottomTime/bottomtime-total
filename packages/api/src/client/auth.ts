import {
  ChangePasswordParamsDTO,
  ChangeRoleParams,
  CurrentUserSchema,
  PasswordResetTokenStatus,
  ResetPasswordParams,
  SuccessFailResponseSchema,
  UserDTO,
  UserRole,
  UserSchema,
  ValidateResetPasswordTokenResponseSchema,
} from '../types';
import { Fetcher } from './fetcher';

export type ChangePasswordResult =
  | {
      succeeded: true;
      user: UserDTO;
    }
  | {
      succeeded: false;
      reason: string;
    };

export class AuthApiClient {
  constructor(private readonly apiClient: Fetcher) {}

  async login(usernameOrEmail: string, password: string): Promise<UserDTO> {
    const { data } = await this.apiClient.post(
      '/api/auth/login',
      {
        usernameOrEmail,
        password,
      },
      UserSchema,
    );
    return data;
  }

  async logout(): Promise<boolean> {
    const {
      data: { succeeded },
    } = await this.apiClient.post(
      '/api/auth/logout',
      undefined,
      SuccessFailResponseSchema,
    );
    return succeeded;
  }

  async getCurrentUser(): Promise<UserDTO | null> {
    const { data } = await this.apiClient.get(
      '/api/auth/me',
      undefined,
      CurrentUserSchema,
    );
    return data.anonymous ? null : data;
  }

  async getOAuthProviders(usernameOrEmail: string): Promise<Set<string>> {
    const { data } = await this.apiClient.get<string[]>(
      `/api/auth/oauth/${usernameOrEmail}`,
    );
    return new Set(data);
  }

  async unlinkOAuthProvider(
    usernameOrEmail: string,
    provider: string,
  ): Promise<void> {
    await this.apiClient.delete(
      `/api/auth/oauth/${usernameOrEmail}/${provider}`,
    );
  }

  async requestPasswordResetToken(usernameOrEmail: string): Promise<void> {
    await this.apiClient.post(
      `/api/users/${usernameOrEmail}/requestPasswordReset`,
    );
  }

  async changePassword(
    user: UserDTO,
    oldPassword: string,
    newPassword: string,
  ): Promise<ChangePasswordResult> {
    const params: ChangePasswordParamsDTO = { oldPassword, newPassword };
    const { data } = await this.apiClient.post(
      `/api/users/${user.username}/password`,
      params,
      SuccessFailResponseSchema,
    );

    if (data.succeeded) {
      return {
        succeeded: true,
        user: {
          ...user,
          hasPassword: true,
          lastPasswordChange: new Date(),
        },
      };
    }

    return {
      succeeded: false,
      reason: data.reason || 'Password change failed.',
    };
  }

  async validatePasswordResetToken(
    usernameOrEmail: string,
    token: string,
  ): Promise<PasswordResetTokenStatus> {
    const { data } = await this.apiClient.get(
      `/api/users/${usernameOrEmail}/resetPassword`,
      { token },
      ValidateResetPasswordTokenResponseSchema,
    );

    return data.status;
  }

  async resetPasswordWithToken(
    usernameOrEmail: string,
    token: string,
    newPassword: string,
  ): Promise<boolean> {
    const { data } = await this.apiClient.post(
      `/api/users/${usernameOrEmail}/resetPassword`,
      {
        newPassword,
        token,
      },
      SuccessFailResponseSchema,
    );

    return data.succeeded;
  }

  async resetPassword(user: UserDTO, newPassword: string): Promise<UserDTO> {
    const params: ResetPasswordParams = { newPassword };
    await this.apiClient.post(
      `/api/admin/users/${user.username}/password`,
      params,
    );
    return {
      ...user,
      hasPassword: true,
      lastPasswordChange: new Date(),
    };
  }

  async changeRole(user: UserDTO, role: UserRole): Promise<UserDTO> {
    const params: ChangeRoleParams = { newRole: role };
    await this.apiClient.post(`/api/admin/users/${user.username}/role`, params);
    return {
      ...user,
      role,
    };
  }

  async toggleAccountLock(user: UserDTO): Promise<UserDTO> {
    const url = `/api/admin/users/${user.username}/${
      user.isLockedOut ? 'unlockAccount' : 'lockAccount'
    }`;
    await this.apiClient.post(url);
    return {
      ...user,
      isLockedOut: !user.isLockedOut,
    };
  }
}
