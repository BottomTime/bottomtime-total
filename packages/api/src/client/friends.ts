import { AxiosInstance, isAxiosError } from 'axios';

import {
  FriendRequestSchema,
  FriendSchema,
  ListFriendRequestsParams,
  ListFriendRequestsResponseSchema,
  ListFriendsParams,
  ListFriendsResposneSchema,
} from '../types';
import { Friend } from './friend';
import { FriendRequest } from './friend-request';

export type ListFriendsResults = {
  friends: Friend[];
  totalCount: number;
};

export type ListFriendRequestsResults = {
  friendRequests: FriendRequest[];
  totalCount: number;
};

export class FriendsApiClient {
  constructor(private readonly apiClient: AxiosInstance) {}

  async listFriends(
    username: string,
    params?: ListFriendsParams,
  ): Promise<ListFriendsResults> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}/friends`,
      { params },
    );
    const result = ListFriendsResposneSchema.parse(data);

    return {
      friends: result.friends.map(
        (friend) => new Friend(this.apiClient, username, friend),
      ),
      totalCount: result.totalCount,
    };
  }

  async areFriends(username: string, friendUsername: string): Promise<boolean> {
    try {
      await this.apiClient.head(
        `/api/users/${username}/friends/${friendUsername}`,
      );
      return true;
    } catch (error) {
      if (isAxiosError(error) && error.response?.status === 404) {
        return false;
      }

      throw error;
    }
  }

  async getFriend(username: string, friendUsername: string): Promise<Friend> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}/friends/${friendUsername}`,
    );
    return new Friend(this.apiClient, username, FriendSchema.parse(data));
  }

  /** @deprecated Use the Friend.unfriend() method instead. Will remove this at some point. */
  async unfriend(username: string, friendUsername: string): Promise<void> {
    await this.apiClient.delete(
      `/api/users/${username}/friends/${friendUsername}`,
    );
  }

  async listFriendRequests(
    username: string,
    params?: ListFriendRequestsParams,
  ): Promise<ListFriendRequestsResults> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}/friendRequests`,
      { params },
    );
    const result = ListFriendRequestsResponseSchema.parse(data);

    return {
      friendRequests: result.friendRequests.map(
        (request) => new FriendRequest(this.apiClient, username, request),
      ),
      totalCount: result.totalCount,
    };
  }

  async createFriendRequest(
    username: string,
    friendUsername: string,
  ): Promise<FriendRequest> {
    const { data } = await this.apiClient.put(
      `/api/users/${username}/friendRequests/${friendUsername}`,
    );

    return new FriendRequest(
      this.apiClient,
      username,
      FriendRequestSchema.parse(data),
    );
  }

  wrapFriendDTO(username: string, data: unknown): Friend {
    const dto = FriendSchema.parse(data);
    return new Friend(this.apiClient, username, dto);
  }

  wrapFriendRequestDTO(username: string, data: unknown): FriendRequest {
    const dto = FriendRequestSchema.parse(data);
    return new FriendRequest(this.apiClient, username, dto);
  }
}
