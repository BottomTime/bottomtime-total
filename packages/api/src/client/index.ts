export * from './errors';

export * from './user-profile';
export * from './user';

export * from './api-client-options';
export * from './fetcher';
export * from './client';
