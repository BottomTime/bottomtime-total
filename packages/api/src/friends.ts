import { SortOrder } from './constants';
import { z } from 'zod';

export enum FriendsSortBy {
  Username = 'username',
  MemberSince = 'memberSince',
  FriendsSince = 'friendsSince',
}

export enum FriendRequestDirection {
  Incoming = 'incoming',
  Outgoing = 'outgoing',
  Both = 'both',
}

// Data Transfer Objects
export const FriendSchema = z.object({
  id: z.string().uuid(),
  friendsSince: z.coerce.date(),
  username: z.string(),
  memberSince: z.coerce.date(),
  avatar: z.string().optional(),
  name: z.string().optional(),
  location: z.string().optional(),
});
export type FriendDTO = z.infer<typeof FriendSchema>;

export const FriendRequestSchema = z.object({
  friendId: z.string().uuid(),
  direction: z.nativeEnum(FriendRequestDirection),
  created: z.coerce.date(),
  expires: z.coerce.date(),
  friend: FriendSchema.omit({ friendsSince: true }),
  accepted: z.boolean().optional(),
  reason: z.string().optional(),
});
export type FriendRequestDTO = z.infer<typeof FriendRequestSchema>;

// Listing Friends
export const ListFriendsParamsSchema = z
  .object({
    sortBy: z.nativeEnum(FriendsSortBy),
    sortOrder: z.nativeEnum(SortOrder),
    skip: z.coerce.number().int().min(0),
    limit: z.coerce.number().int().min(1).max(100),
  })
  .partial();
export type ListFriendsParams = z.infer<typeof ListFriendsParamsSchema>;

export const ListFriendsResposneSchema = z.object({
  friends: FriendSchema.array(),
  totalCount: z.number().int(),
});
export type ListFriendsResponseDTO = z.infer<typeof ListFriendsResposneSchema>;

// Listing Friend Requests
export const ListFriendRequestsParamsSchema = z
  .object({
    direction: z
      .nativeEnum(FriendRequestDirection)
      .default(FriendRequestDirection.Both),
    skip: z.coerce.number().int().min(0).default(0),
    limit: z.coerce.number().int().min(1).max(200).default(50),
  })
  .partial();
export type ListFriendRequestsParams = z.infer<
  typeof ListFriendRequestsParamsSchema
>;

export const ListFriendRequestsResponseSchema = z.object({
  friendRequests: FriendRequestSchema.array(),
  totalCount: z.number().int(),
});
export type ListFriendRequestsResponseDTO = z.infer<
  typeof ListFriendRequestsResponseSchema
>;

// Acknowledging Friend Requests
export const AcknowledgeFriendRequestParamsSchema = z.discriminatedUnion(
  'accepted',
  [
    z.object({
      accepted: z.literal(true),
    }),
    z.object({
      accepted: z.literal(false),
      reason: z.string().optional(),
    }),
  ],
);
export type AcknowledgeFriendRequestParamsDTO = z.infer<
  typeof AcknowledgeFriendRequestParamsSchema
>;
