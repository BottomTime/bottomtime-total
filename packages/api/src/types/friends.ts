import { z } from 'zod';

import { BooleanString, SortOrder } from './constants';
import { LogBookSharing } from './users';

export enum FriendsSortBy {
  Username = 'username',
  MemberSince = 'memberSince',
  FriendsSince = 'friendsSince',
}

export enum FriendRequestDirection {
  Incoming = 'incoming',
  Outgoing = 'outgoing',
  Both = 'both',
}

// Data Transfer Objects
export const FriendSchema = z.object({
  id: z.string().uuid(),
  friendsSince: z.coerce.date(),
  username: z.string(),
  memberSince: z.coerce.date(),
  logBookSharing: z.nativeEnum(LogBookSharing),
  avatar: z.string().optional(),
  name: z.string().optional(),
  location: z.string().optional(),
});
export type FriendDTO = z.infer<typeof FriendSchema>;

export const FriendRequestSchema = z.object({
  friendId: z.string().uuid(),
  direction: z.nativeEnum(FriendRequestDirection),
  created: z.coerce.date(),
  expires: z.coerce.date(),
  friend: FriendSchema.omit({ friendsSince: true }),
  accepted: z.boolean().optional(),
  reason: z.string().optional(),
});
export type FriendRequestDTO = z.infer<typeof FriendRequestSchema>;

// Listing Friends
export const ListFriendsParamsSchema = z
  .object({
    sortBy: z.nativeEnum(FriendsSortBy),
    sortOrder: z.nativeEnum(SortOrder),
    skip: z.coerce.number().int().min(0),
    limit: z.coerce.number().int().min(1).max(100),
  })
  .partial();
export type ListFriendsParamsDTO = z.infer<typeof ListFriendsParamsSchema>;

export const ListFriendsResposneSchema = z.object({
  data: FriendSchema.array(),
  totalCount: z.number().int(),
});

// Listing Friend Requests
export const ListFriendRequestsParamsSchema = z
  .object({
    direction: z.nativeEnum(FriendRequestDirection),
    showAcknowledged: BooleanString,
    showExpired: BooleanString,
    skip: z.coerce.number().int().min(0),
    limit: z.coerce.number().int().min(1).max(200),
  })
  .partial();
export type ListFriendRequestsParamsDTO = z.infer<
  typeof ListFriendRequestsParamsSchema
>;

export const ListFriendRequestsResponseSchema = z.object({
  data: FriendRequestSchema.array(),
  totalCount: z.number().int(),
});

// Acknowledging Friend Requests
export const AcknowledgeFriendRequestParamsSchema = z.discriminatedUnion(
  'accepted',
  [
    z.object({
      accepted: z.literal(true),
    }),
    z.object({
      accepted: z.literal(false),
      reason: z.string().max(500).optional(),
    }),
  ],
);
export type AcknowledgeFriendRequestParamsDTO = z.infer<
  typeof AcknowledgeFriendRequestParamsSchema
>;
