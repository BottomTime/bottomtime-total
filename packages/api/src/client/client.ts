import { AppMetricsDTO } from '../types';
import { AlertsApiClient } from './alerts';
import { ApiClientOptions } from './api-client-options';
import { AuthApiClient } from './auth';
import { CertificationsApiClient } from './certifications';
import { DiveSiteReviewsApiClient } from './dive-site-reviews';
import { DiveSitesApiClient } from './dive-sites';
import { Fetcher } from './fetcher';
import { FriendsApiClient } from './friends';
import { LogEntriesApiClient } from './log-entries';
import { MembershipsApiClient } from './memberships';
import { NotificationsApiClient } from './notifications';
import { OperatorReviewsApiClient } from './operator-reviews';
import { OperatorsApiClient } from './operators';
import { TanksApiClient } from './tanks';
import { UserAccountsApiClient } from './user-accounts';
import { UserProfilesApiClient } from './user-profiles';

export class ApiClient {
  private readonly client: Fetcher;

  readonly alerts: AlertsApiClient;
  readonly auth: AuthApiClient;
  readonly certifications: CertificationsApiClient;
  readonly diveSiteReviews: DiveSiteReviewsApiClient;
  readonly diveSites: DiveSitesApiClient;
  readonly friends: FriendsApiClient;
  readonly logEntries: LogEntriesApiClient;
  readonly memberships: MembershipsApiClient;
  readonly notifications: NotificationsApiClient;
  readonly operatorReviews: OperatorReviewsApiClient;
  readonly operators: OperatorsApiClient;
  readonly tanks: TanksApiClient;
  readonly userAccounts: UserAccountsApiClient;
  readonly userProfiles: UserProfilesApiClient;

  constructor(options?: ApiClientOptions) {
    this.client = options?.fetcher ?? new Fetcher(options);

    this.alerts = new AlertsApiClient(this.client);
    this.auth = new AuthApiClient(this.client);
    this.certifications = new CertificationsApiClient(this.client);
    this.diveSiteReviews = new DiveSiteReviewsApiClient(this.client);
    this.diveSites = new DiveSitesApiClient(this.client);
    this.friends = new FriendsApiClient(this.client);
    this.logEntries = new LogEntriesApiClient(this.client);
    this.memberships = new MembershipsApiClient(this.client);
    this.notifications = new NotificationsApiClient(this.client);
    this.operatorReviews = new OperatorReviewsApiClient(this.client);
    this.operators = new OperatorsApiClient(this.client);
    this.tanks = new TanksApiClient(this.client);
    this.userAccounts = new UserAccountsApiClient(this.client);
    this.userProfiles = new UserProfilesApiClient(this.client);
  }

  async getAppMetrics(): Promise<AppMetricsDTO> {
    const { data } = await this.client.get<AppMetricsDTO>('/api/metrics');
    return data;
  }
}
