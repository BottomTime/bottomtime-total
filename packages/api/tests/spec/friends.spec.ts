import axios, { AxiosInstance } from 'axios';
import nock, { Scope } from 'nock';

import {
  Friend,
  FriendRequest,
  FriendRequestDirection,
  FriendsSortBy,
  ListFriendRequestsParams,
  ListFriendRequestsResponseDTO,
  ListFriendRequestsResponseSchema,
  ListFriendsParams,
  ListFriendsResponseDTO,
  ListFriendsResposneSchema,
  SortOrder,
} from '../../src';
import { FriendsApiClient } from '../../src/client/friends';
import FriendRequestTestData from '../fixtures/friend-requests-search-results.json';
import FriendTestData from '../fixtures/friends-search-results.json';
import { createScope } from '../fixtures/nock';

const Username = 'mega_user32';

describe('Friends API client', () => {
  let axiosClient: AxiosInstance;
  let client: FriendsApiClient;
  let scope: Scope;
  let friendsData: ListFriendsResponseDTO;
  let friendRequestData: ListFriendRequestsResponseDTO;

  beforeAll(() => {
    axiosClient = axios.create();
    scope = createScope();
    client = new FriendsApiClient(axiosClient);

    friendsData = ListFriendsResposneSchema.parse(FriendTestData);
    friendRequestData = ListFriendRequestsResponseSchema.parse(
      FriendRequestTestData,
    );
  });

  afterEach(() => {
    nock.cleanAll();
  });

  afterAll(() => {
    nock.restore();
  });

  it('will perform basic search for friends', async () => {
    scope.get(`/api/users/${Username}/friends`).reply(200, friendsData);
    const result = await client.listFriends(Username);

    expect(scope.isDone()).toBe(true);
    expect(result.totalCount).toBe(friendsData.totalCount);
    expect(result.friends).toHaveLength(friendsData.friends.length);
    result.friends.forEach((friend, index) => {
      expect(friend.id).toBe(friendsData.friends[index].id);
      expect(friend.username).toBe(friendsData.friends[index].username);
      expect(friend.avatar).toBe(friendsData.friends[index].avatar);
      expect(friend.location).toBe(friendsData.friends[index].location);
      expect(friend.name).toBe(friendsData.friends[index].name);
      expect(friend.memberSince).toEqual(
        friendsData.friends[index].memberSince,
      );
      expect(friend.logBookSharing).toBe(
        friendsData.friends[index].logBookSharing,
      );
    });
  });

  it('will perform a search for friends with parameters', async () => {
    const options: ListFriendsParams = {
      sortBy: FriendsSortBy.Username,
      sortOrder: SortOrder.Ascending,
      skip: 20,
      limit: 10,
    };

    scope
      .get(`/api/users/${Username}/friends`)
      .query(options)
      .reply(200, friendsData);

    const result = await client.listFriends(Username, options);

    expect(scope.isDone()).toBe(true);
    expect(result.totalCount).toBe(friendsData.totalCount);
    expect(result.friends).toHaveLength(friendsData.friends.length);
    result.friends.forEach((friend, index) => {
      expect(friend.id).toBe(friendsData.friends[index].id);
      expect(friend.username).toBe(friendsData.friends[index].username);
      expect(friend.avatar).toBe(friendsData.friends[index].avatar);
      expect(friend.location).toBe(friendsData.friends[index].location);
      expect(friend.name).toBe(friendsData.friends[index].name);
      expect(friend.memberSince).toEqual(
        friendsData.friends[index].memberSince,
      );
      expect(friend.logBookSharing).toBe(
        friendsData.friends[index].logBookSharing,
      );
    });
  });

  it('will return a single friend', async () => {
    const data = friendsData.friends[0];
    scope
      .get(`/api/users/${Username}/friends/${data.username}`)
      .reply(200, data);

    const friend = await client.getFriend(Username, data.username);

    expect(scope.isDone()).toBe(true);
    expect(friend.toJSON()).toEqual(data);
  });

  it('will return true if areFriends is called and the two users are friends', async () => {
    const friend = 'my_homie';
    scope.head(`/api/users/${Username}/friends/${friend}`).reply(200);
    await expect(client.areFriends(Username, friend)).resolves.toBe(true);
    expect(scope.isDone()).toBe(true);
  });

  it('will return false if areFriends is called and the two users are not friends', async () => {
    const friend = 'my_homie';
    scope.head(`/api/users/${Username}/friends/${friend}`).reply(404);
    await expect(client.areFriends(Username, friend)).resolves.toBe(false);
    expect(scope.isDone()).toBe(true);
  });

  it('will allow an error to bubble up if areFriends is called and something goes wrong', async () => {
    const friend = 'my_homie';
    const error = new Error('nope');
    scope
      .head(`/api/users/${Username}/friends/${friend}`)
      .replyWithError(error);
    await expect(
      client.areFriends(Username, friend),
    ).rejects.toThrowErrorMatchingSnapshot();
    expect(scope.isDone()).toBe(true);
  });

  it('will perform a basic search for friend requests', async () => {
    scope
      .get(`/api/users/${Username}/friendRequests`)
      .reply(200, friendRequestData);

    const result = await client.listFriendRequests(Username);

    expect(scope.isDone()).toBe(true);
    expect(result.totalCount).toBe(friendRequestData.totalCount);
    expect(result.friendRequests).toHaveLength(
      friendRequestData.friendRequests.length,
    );
    result.friendRequests.forEach((request, index) => {
      expect(request.toJSON()).toEqual(friendRequestData.friendRequests[index]);
    });
  });

  it('will perform a search for friend requests with parameters', async () => {
    const options: ListFriendRequestsParams = {
      direction: FriendRequestDirection.Incoming,
      showAcknowledged: true,
      showExpired: true,
      skip: 20,
      limit: 10,
    };

    scope
      .get(`/api/users/${Username}/friendRequests`)
      .query(options)
      .reply(200, friendRequestData);

    const result = await client.listFriendRequests(Username, options);

    expect(scope.isDone()).toBe(true);
    expect(result.totalCount).toBe(friendRequestData.totalCount);
    expect(result.friendRequests).toHaveLength(
      friendRequestData.friendRequests.length,
    );
    result.friendRequests.forEach((request, index) => {
      expect(request.toJSON()).toEqual(friendRequestData.friendRequests[index]);
    });
  });

  it('will create a friend request', async () => {
    const expected = friendRequestData.friendRequests[0];
    scope
      .put(`/api/users/${Username}/friendRequests/${expected.friend.username}`)
      .reply(201, expected);

    const request = await client.createFriendRequest(
      Username,
      expected.friend.username,
    );

    expect(scope.isDone()).toBe(true);
    expect(request.toJSON()).toEqual(expected);
  });

  it('will wrap a friend DTO', () => {
    const data = friendsData.friends[0];
    const friend = client.wrapFriendDTO(Username, data);

    expect(friend).toBeInstanceOf(Friend);
    expect(friend.toJSON()).toEqual(data);
  });

  it('will wrap a friend request DTO', () => {
    const data = friendRequestData.friendRequests[0];
    const request = client.wrapFriendRequestDTO(Username, data);

    expect(request).toBeInstanceOf(FriendRequest);
    expect(request.toJSON()).toEqual(data);
  });
});
