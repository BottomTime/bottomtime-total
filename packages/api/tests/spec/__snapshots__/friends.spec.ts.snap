// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends API client will allow an error to bubble up if areFriends is called and something goes wrong 1`] = `"nope"`;
